### PR TITLE
Add export option copy to clipboard in Markdown format

### PR DESCRIPTION
**Pull request by timengel**
## Changelog

- [x] Added option to export board data to clipboard in a compact Markdown format

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] The light- and dark-theme are both supported and tested
- [x] The design was implemented and is responsive for all devices and screen sizes
- [x] The application was tested in the most commonly used browsers (e.g. Chrome, Firefox, Safari)

### CODE REVIEW

**Comment by timengel** [comment]
Example of MD export:  
<hr />

# Lorem ipsum
## Columns
### Positive (12 Notes)

- qwdasdqweqwe (Odd Orc, 9 Votes)
- asdasdasdasdasd (Odd Orc, 5 Votes)
    - sdf (Odd Orc, 3 Votes)
    - sdf (Odd Orc)
- swdfsdf (Odd Orc)
- Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet  (Odd Orc)
- Lorem ipsum dolor sit amet, consetetur sadipscing (Odd Orc)
- Lorem ipsum dolor sit amet (Odd Orc)
- Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet 3 Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet 3 Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet 3 Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet 3 Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet (Odd Orc)
    - Lorem ipsum dolor sit amet (Odd Orc)
    - Lorem ipsum dolor sit ametLorem ipsum dolor sit amet (Odd Orc)
    - Lorem ipsum dolor sit amet (Odd Orc)

### Negative (8 Notes)

- sdf jhughdh (Odd Orc, 5 Votes)
- sfd asdf (Odd Orc, 2 Votes)
- Lorem ipsum dolor sit amet (Odd Orc, 2 Votes)
    - Lorem ipsum dolor sit amet (Odd Orc)
- sdf (Odd Orc, 1 Vote)
- Lorem ipsum dolor sit ametLorem ipsum dolor sit amet (Odd Orc)
- Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet (Odd Orc)
- Lorem ipsum dolor sit amet (Odd Orc)

### Lorem ipsum dolor sit amet (8 Notes)

- Lorem ipsum dolor sit amet (Odd Orc, 3 Votes)
- dfs Lorem ipsum dolor sit amet (Odd Orc)
    - 5 Lorem ipsum dolor sit amet (Odd Orc)
    - s ASffdd (Odd Orc)
- Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet (Odd Orc)
    - Lorem ipsum dolor sit ametLorem sum dolor sit amet (Odd Orc)
    - Lorem ipsum dolor sit amet (Odd Orc)
- Lorem ipsum dolor sit amet (Odd Orc)

### Qwertz (3 Notes)

- asdasdasd (Odd Orc)
- Bacon ipsum dolor amet kielbasa rump ham buffalo jowl frankfurter meatloaf bacon tongue cow corned beef pork venison.  (Furious Frog)
- Bacon ipsum dolor amet kielbasa rump ham buffalo jowl frankfurter meatloaf bacon tongue cow corned beef pork venison.  (Furious Frog)

## Board Properties
- Board-ID: 15d21cf5-94da-4afc-86bf-5d6412887f5f
- Participants: 5

> Generated on [scrumlr.io](https://www.scrumlr.io)  
Provided with ♥️ by [inovex](https://www.inovex.de)  

Scrumlr Logo

File: src/utils/export.ts [hunk 1]
```diff
@@ -18,3 +20,93 @@ export const exportAsJSON = async (id: string, name?: string) => {
...
+  return `${header + columnList}\n\n`;
+};
+
+const mdBranding = () =>
+  `> ${i18n.t("PrintView.GeneratedOn")} [scrumlr.io](https://www.scrumlr.io)  \n${t("PrintView.ProvidedBy")} [inovex](https://www.inovex.de)  \n\n![Scrumlr Logo](${
+    process.env.REACT_APP_PUBLIC_URL
+  }/scrumlr-logo-light.svg)`;
+
+const mdTemplate = (boardData: BoardDataType) =>
+  mdBoardHeader(boardData.board.name) + mdColumns(boardData) + mdBoardProperties(boardData.board, boardData.participants) + mdBranding();
```
Comment (bitionaire) [inline src/utils/export.ts:106]:
```suggestion
  mdBoardHeader(boardData.board.name || "scrumlr.io") + mdBoardProperties(boardData.board, boardData.participants) + mdColumns(boardData) + mdBranding();
```

File: src/utils/export.ts [hunk 2]
```diff
@@ -18,3 +20,93 @@ export const exportAsJSON = async (id: string, name?: string) => {
...
+  return a.position.rank > b.position.rank ? -1 : 1;
+};
+
+export const getChildNotes = (notes: NoteType[], votings: VotingType[], noteId: string) =>
+  notes.filter((n) => n.position.stack === noteId).sort((a, b) => compareNotes(a, b, votings));
+
+export const getAuthorName = (authorId: string, participants: ParticipantType[]) => participants.filter((p: ParticipantType) => p.user?.id === authorId)[0].user?.name;
+
+const {t} = i18n;
+const mdBoardHeader = (boardName: string) => `# ${boardName}\n`;
```
Comment (bitionaire) [inline src/utils/export.ts:39]:
```suggestion
const mdBoardHeader = (boardName: string) => `# ${boardName}\n\n`;
```

File: src/utils/export.ts [hunk 3]
```diff
@@ -18,3 +20,93 @@ export const exportAsJSON = async (id: string, name?: string) => {
...
+    .map((n) => {
+      const votes = boardData.board.showVoting ? mdVotesPerNote(n.id, boardData.votings) : "";
+      const author = boardData.board.showAuthors ? getAuthorName(n.author, boardData.participants) : "";
+      const stack = mdStack(getChildNotes(boardData.notes, boardData.votings, n.id), boardData);
+      return mdNote(n, author, votes, stack);
+    })
+    .join("");
+
+const mdColumns = (boardData: BoardDataType) => {
+  const header = `## ${t("MarkdownExport.columns")}\n`;
```
Comment (bitionaire) [inline src/utils/export.ts:90]:
```suggestion
```

File: src/utils/export.ts [hunk 4]
```diff
@@ -18,3 +20,93 @@ export const exportAsJSON = async (id: string, name?: string) => {
...
+  const columnList = boardData.columns
+    .filter((c) => boardData.notes.filter((n) => n.position.column === c.id).length > 0)
+    .map(
+      (c: ColumnType) => `### ${c.name} (${boardData.notes.filter((n) => n.position.column === c.id).length} ${t("MarkdownExport.notes")})\n${mdNotesPerColumn(c.id, boardData)}`
+    )
+    .join("\n\n");
+  return `${header + columnList}\n\n`;
```
Comment (bitionaire) [inline src/utils/export.ts:97]:
```suggestion
  return `${columnList}\n\n`;
```

File: src/utils/export.ts [hunk 5]
```diff
@@ -18,3 +20,93 @@ export const exportAsJSON = async (id: string, name?: string) => {
...
+      return mdNote(n, author, votes, stack);
+    })
+    .join("");
+
+const mdColumns = (boardData: BoardDataType) => {
+  const header = `## ${t("MarkdownExport.columns")}\n`;
+  const columnList = boardData.columns
+    .filter((c) => boardData.notes.filter((n) => n.position.column === c.id).length > 0)
+    .map(
+      (c: ColumnType) => `### ${c.name} (${boardData.notes.filter((n) => n.position.column === c.id).length} ${t("MarkdownExport.notes")})\n${mdNotesPerColumn(c.id, boardData)}`
```
Comment (bitionaire) [inline src/utils/export.ts:94]:
```suggestion
      (c: ColumnType) => `## ${c.name} (${boardData.notes.filter((n) => n.position.column === c.id).length} ${t("MarkdownExport.notes")})\n${mdNotesPerColumn(c.id, boardData)}`
```

File: src/utils/export.ts [hunk 6]
```diff
@@ -18,3 +20,93 @@ export const exportAsJSON = async (id: string, name?: string) => {
...
+  const authorsHidden = !board.showAuthors ? `- ${t("MarkdownExport.showAuthorsDisabled")}\n` : "";
+  return `${header + boardId + participantsNumber + authorsHidden}\n`;
+};
+
+const mdVotesPerNote = (noteId: string, votings: VotingType[]) => {
+  const votes = votings ? getNoteVotes(noteId, votings) : 0;
+  return votes ? `${votes} ${votes === 1 ? t("MarkdownExport.vote") : t("MarkdownExport.votes")}` : "";
+};
+
+const mdBrackets = (addBrackets: boolean, input: string) => (addBrackets ? `(${input})` : input);
```
Comment (bitionaire) [inline src/utils/export.ts:54]:
```suggestion
const mdBrackets = (addBrackets: boolean, input: string) => (addBrackets ? `_(${input})_` : input);
```

File: src/utils/export.ts [hunk 7]
```diff
@@ -18,3 +20,93 @@ export const exportAsJSON = async (id: string, name?: string) => {
...
+    .map(
+      (c: ColumnType) => `### ${c.name} (${boardData.notes.filter((n) => n.position.column === c.id).length} ${t("MarkdownExport.notes")})\n${mdNotesPerColumn(c.id, boardData)}`
+    )
+    .join("\n\n");
+  return `${header + columnList}\n\n`;
+};
+
+const mdBranding = () =>
+  `> ${i18n.t("PrintView.GeneratedOn")} [scrumlr.io](https://www.scrumlr.io)  \n${t("PrintView.ProvidedBy")} [inovex](https://www.inovex.de)  \n\n![Scrumlr Logo](${
+    process.env.REACT_APP_PUBLIC_URL
```
Comment (bitionaire) [inline src/utils/export.ts:102]:
```suggestion
    process.env.REACT_APP_PUBLIC_URL || "https://scrumlr.io"
```

File: src/utils/export.ts [hunk 8]
```diff
@@ -18,3 +20,93 @@ export const exportAsJSON = async (id: string, name?: string) => {
...
+export const getChildNotes = (notes: NoteType[], votings: VotingType[], noteId: string) =>
+  notes.filter((n) => n.position.stack === noteId).sort((a, b) => compareNotes(a, b, votings));
+
+export const getAuthorName = (authorId: string, participants: ParticipantType[]) => participants.filter((p: ParticipantType) => p.user?.id === authorId)[0].user?.name;
+
+const {t} = i18n;
+const mdBoardHeader = (boardName: string) => `# ${boardName}\n`;
+
+const mdBoardProperties = (board: BoardType, participants: ParticipantType[]) => {
+  const header = `## ${i18n.t("MarkdownExport.properties")}\n`;
```
Comment (bitionaire) [inline src/utils/export.ts:42]:
```suggestion
```

File: src/utils/export.ts [hunk 9]
```diff
@@ -18,3 +20,93 @@ export const exportAsJSON = async (id: string, name?: string) => {
...
+  const boardId = `- ${i18n.t("MarkdownExport.boardId")}: ${board.id}\n`;
+  const participantsNumber = `- ${t("MarkdownExport.participants")}: ${participants.length}\n`;
+  const authorsHidden = !board.showAuthors ? `- ${t("MarkdownExport.showAuthorsDisabled")}\n` : "";
+  return `${header + boardId + participantsNumber + authorsHidden}\n`;
```
Comment (bitionaire) [inline src/utils/export.ts:46]:
```suggestion
  return `${boardId + participantsNumber + authorsHidden}\n`;
```

File: public/locales/en/translation.json [hunk 10]
```diff
@@ -81,6 +82,18 @@
     "Notes": "Kärtchen",
     "Votes": "Stimmen"
   },
+  "MarkdownExport": {
+    "properties": "Board Properties",
```
Comment (bitionaire) [inline public/locales/en/translation.json:86]:
```suggestion
```

File: public/locales/en/translation.json [hunk 11]
```diff
@@ -81,6 +82,18 @@
...
+    "columns": "Columns",
+    "note": "Note",
+    "notes": "Notes",
+    "participants": "Participants",
+    "boardId": "Board-ID",
```
Comment (bitionaire) [inline public/locales/en/translation.json:91]:
```suggestion
    "boardId": "Link zum Board",
```

File: public/locales/de/translation.json [hunk 12]
```diff
@@ -72,14 +72,27 @@
...
   "PrintView": {
     "Participants": "Teilnehmer",
     "GeneratedOn": "Erstellt auf",
     "ProvidedBy": "Bereitgestellt mit ♥️ von",
     "Notes": "Kärtchen",
-    "Votes": "Votes"
+    "Votes": "Stimmen"
+  },
+  "MarkdownExport": {
+    "properties": "Sitzungseigenschaften",
```
Comment (bitionaire) [inline public/locales/de/translation.json:86]:
```suggestion
```

File: public/locales/de/translation.json [hunk 13]
```diff
@@ -72,14 +72,27 @@
...
+    "columns": "Spalten",
+    "note": "Kärtchen",
+    "notes": "Kärtchen",
+    "participants": "Teilnehmer",
+    "boardId": "Sitzungs-ID",
```
Comment (bitionaire) [inline public/locales/de/translation.json:91]:
```suggestion
    "boardId": "Link to board",
```

File: src/utils/export.ts [hunk 14]
```diff
@@ -18,3 +20,93 @@ export const exportAsJSON = async (id: string, name?: string) => {
...
+  notes.filter((n) => n.position.stack === noteId).sort((a, b) => compareNotes(a, b, votings));
+
+export const getAuthorName = (authorId: string, participants: ParticipantType[]) => participants.filter((p: ParticipantType) => p.user?.id === authorId)[0].user?.name;
+
+const {t} = i18n;
+const mdBoardHeader = (boardName: string) => `# ${boardName}\n`;
+
+const mdBoardProperties = (board: BoardType, participants: ParticipantType[]) => {
+  const header = `## ${i18n.t("MarkdownExport.properties")}\n`;
+  const boardId = `- ${i18n.t("MarkdownExport.boardId")}: ${board.id}\n`;
```
Comment (bitionaire) [inline src/utils/export.ts:43]:
```suggestion
  const boardId = `- [${i18n.t("MarkdownExport.boardId")}](${window.location.href.replace("/settings/export", "")})\n`;
```

The string replacement is not the best option. That's just quick and dirty to include the link without the necessity to use PUBLIC_URL

**Review by bitionaire — CHANGES_REQUESTED** [review]
Would be awesome if you could also add a unit test for the export function. Seems pretty straightforward and will help us in the future :)

File: src/utils/export.ts (hunk 14 above)
Comment (timengel) [reply src/utils/export.ts:43]: What is wrong with using the environment variable/`PUBLIC_URL`?

File: src/utils/export.ts (hunk 7 above)
Comment (timengel) [reply src/utils/export.ts:102]: I used a constant `DEFAULT_URL` instead :)

File: src/utils/export.ts (hunk 1 above)
Comment (timengel) [reply src/utils/export.ts:106]: I used the existing constant `DEFAULT_BOARD_NAME` instead :)

File: src/utils/export.ts (hunk 6 above)
Comment (timengel) [reply src/utils/export.ts:55]: I changed the function name to `mdItalicBrackets`

File: src/utils/export.ts (hunk 14 above)
Comment (bitionaire) [reply src/utils/export.ts]: You have to manually set it and it's easy to forget about that

**Comment by timengel** [comment]
@bitionaire can this be merged?

**Review by bitionaire — APPROVED** [review]

**Pull request merged**